### PR TITLE
modularize rustdoc's write_shared

### PR DESCRIPTION
**Pull request by EtomicBomb**
Refactor src/librustdoc/html/render/write_shared.rs to reduce code duplication, adding unit tests

* Extract + unit test code for sorting and rendering JSON, which is duplicated 9 times in the current impl
* Extract + unit test code for encoding JSON as single quoted strings, which is duplicated twice in the current impl
* Unit tests for cross-crate information file formats
* Generic interface to add new kinds of cross-crate information files in the future
* Intended to match current behavior exactly, except for a merge info comment it adds to the bottom of cci files
* This PR is intended to reduce the review burden from my [mergeable rustdoc rfc](https://github.com/rust-lang/rfcs/pull/3662) implementation PR, which is a [small commit based on this branch](https://github.com/EtomicBomb/rust/tree/rfc). This code is agnostic to the RFC and does not include any of the flags discussed there, but cleanly enables the addition of these flags in the future because it is more modular

### CODE REVIEW

**Comment by EtomicBomb** [comment]
Made a mistake when rebasing, just force-pushed a fix. There are no changes to compiletest.

File: src/librustdoc/html/render/ordered_json.rs [hunk 1]
```diff
@@ -0,0 +1,87 @@
+use std::borrow::Borrow;
+use std::fmt;
+
+use itertools::Itertools as _;
+use serde::{Deserialize, Serialize};
+use serde_json::Value;
+
+/// Prerenedered json.
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/ordered_json.rs:8]:
```suggestion
/// Prerendered json.
```

File: src/librustdoc/html/render/ordered_json.rs [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+/// Both the Display and serde_json::to_string implementations write the serialized json
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord, Serialize, Deserialize)]
+#[serde(from = "Value")]
+#[serde(into = "Value")]
+pub(crate) struct OrderedJson(String);
+
+impl OrderedJson {
+    /// If you pass in an array, it will not be sorted.
+    pub(crate) fn serialize<T: Serialize>(item: T) -> Result<Self, serde_json::Error> {
+        Ok(OrderedJson(serde_json::to_string(&item)?))
```
Comment (GuillaumeGomez) [inline src/librustdoc/html/render/ordered_json.rs:19]: Please use `Self` as much as possible. It makes it easier if the type name is updated later on.

**Comment by GuillaumeGomez** [comment]
Lot of things are going on here. Since it's updating the search index, cc @notriddle 

Also, I'm curious about the perf impact.

@bors try @rust-timer queue

File: src/librustdoc/html/render/ordered_json.rs (hunk 2 above)
Comment (EtomicBomb) [reply src/librustdoc/html/render/ordered_json.rs]: Thanks for taking a look Guillaume!

**Pull request merged**